File: src/pkg/schemas/knowledge.py
from datetime import datetime

from pydantic import BaseModel


class SaveDocumentRequest(BaseModel):
    message_content: str
    title: str | None = None
    storage_uri: str | None = None
    document_format: str | None = None
    document_filename: str | None = None
    session_id: str | None = None
    category_id: int | None = None


class SaveDocumentResponse(BaseModel):
    source_id: str
    note_id: str


class RememberRequest(BaseModel):
    content: str
    session_id: str | None = None
    title: str | None = None
    category_id: int | None = None


class KnowledgeStatsRead(BaseModel):
    item_id: str
    item_type: str
    title: str
    category_name: str | None = None
    search_count: int
    retrieval_count: int
    reference_count: int
    total_count: int
    last_accessed_at: datetime | None = None


class KnowledgeStatsList(BaseModel):
    items: list[KnowledgeStatsRead]
    total: int


class DashboardCounts(BaseModel):
    notes: int
    sources: int
    chats: int
    digest_pending: int


class DashboardTrendDay(BaseModel):
    date: str
    notes: int = 0
    sources: int = 0
    chats: int = 0


class CategoryDistribution(BaseModel):
    name: str
    display_name: str
    notes: int = 0
    sources: int = 0


class NoteTypeDistribution(BaseModel):
    type: str
    count: int


class DashboardResponse(BaseModel):
    counts: DashboardCounts
    trends: list[DashboardTrendDay]
    category_distribution: list[CategoryDistribution]
    note_type_distribution: list[NoteTypeDistribution]

File: web/src/lib/api/knowledge.ts
import { request } from "./client";
import type { Note } from "./notes";

export interface KnowledgeStatsItem {
  item_id: string;
  item_type: string;
  title: string;
  category_name?: string | null;
  search_count: number;
  retrieval_count: number;
  reference_count: number;
  total_count: number;
  last_accessed_at?: string | null;
}

export interface KnowledgeStatsList {
  items: KnowledgeStatsItem[];
  total: number;
}

export interface SaveDocumentRequest {
  message_content: string;
  title?: string;
  storage_uri?: string;
  document_format?: string;
  document_filename?: string;
  session_id?: string;
  category_id?: number;
}

export interface SaveDocumentResponse {
  source_id: string;
  note_id: string;
}

export interface RememberRequest {
  content: string;
  session_id?: string;
  title?: string;
}

export interface ModelInfo {
  id: string;
  display_name: string;
  provider_id: string;
  provider_name: string;
}

export interface DashboardData {
  counts: {
    notes: number;
    sources: number;
    chats: number;
    digest_pending: number;
  };
  trends: Array<{
    date: string;
    notes: number;
    sources: number;
    chats: number;
  }>;
  category_distribution: Array<{
    name: string;
    display_name: string;
    notes: number;
    sources: number;
  }>;
  note_type_distribution: Array<{
    type: string;
    count: number;
  }>;
}

export const knowledgeApi = {
  models: () => request<ModelInfo[]>("/knowledge/models"),
  dashboard: () => request<DashboardData>("/knowledge/dashboard"),
  saveDocument: (body: SaveDocumentRequest) =>
    request<SaveDocumentResponse>("/knowledge/save-document", {
      method: "POST",
      body: JSON.stringify(body),
    }),
  remember: (body: RememberRequest) =>
    request<Note>("/knowledge/remember", {
      method: "POST",
      body: JSON.stringify(body),
    }),
  stats: (params?: {
    sort_by?: string;
    item_type?: string;
    title?: string;
    limit?: number;
    offset?: number;
  }) => {
    const q = new URLSearchParams();
    if (params?.sort_by) q.set("sort_by", params.sort_by);
    if (params?.item_type) q.set("item_type", params.item_type);
    if (params?.title) q.set("title", params.title);
    if (params?.limit) q.set("limit", String(params.limit));
    if (params?.offset) q.set("offset", String(params.offset));
    return request<KnowledgeStatsList>(`/knowledge/stats?${q}`);
  },
};
